File: cloudflare/worker/src/language.mjs
/**
 * Detect the dominant language of user seeds via Unicode script analysis.
 */

function scriptOf(ch) {
  const code = ch.codePointAt(0);

  // CJK Unified Ideographs
  if (
    (code >= 0x4e00 && code <= 0x9fff) ||
    (code >= 0x3400 && code <= 0x4dbf) ||
    (code >= 0x20000 && code <= 0x2a6df) ||
    (code >= 0xf900 && code <= 0xfaff)
  ) {
    return "CJK";
  }

  // Hiragana
  if (code >= 0x3040 && code <= 0x309f) return "HIRAGANA";

  // Katakana
  if (
    (code >= 0x30a0 && code <= 0x30ff) ||
    (code >= 0x31f0 && code <= 0x31ff)
  ) {
    return "KATAKANA";
  }

  // Hangul
  if (
    (code >= 0xac00 && code <= 0xd7af) ||
    (code >= 0x1100 && code <= 0x11ff) ||
    (code >= 0x3130 && code <= 0x318f)
  ) {
    return "HANGUL";
  }

  // Cyrillic
  if (
    (code >= 0x0400 && code <= 0x04ff) ||
    (code >= 0x0500 && code <= 0x052f)
  ) {
    return "CYRILLIC";
  }

  // Arabic
  if (
    (code >= 0x0600 && code <= 0x06ff) ||
    (code >= 0x0750 && code <= 0x077f) ||
    (code >= 0x08a0 && code <= 0x08ff)
  ) {
    return "ARABIC";
  }

  // Thai
  if (code >= 0x0e00 && code <= 0x0e7f) return "THAI";

  // Devanagari
  if (code >= 0x0900 && code <= 0x097f) return "DEVANAGARI";

  return "OTHER";
}

export function detectSeedLanguage(seeds) {
  const combined = seeds.join(" ").trim();
  if (!combined) return null;

  const counts = {};
  let totalNonWs = 0;

  for (const ch of combined) {
    if (/\s/.test(ch)) continue;
    totalNonWs++;
    const script = scriptOf(ch);
    counts[script] = (counts[script] || 0) + 1;
  }

  if (totalNonWs === 0) return null;

  const threshold = 0.30;
  const kana = (counts.HIRAGANA || 0) + (counts.KATAKANA || 0);
  const cjk = counts.CJK || 0;
  const hangul = counts.HANGUL || 0;

  // Japanese: any Kana present + CJK
  if (kana > 0 && (kana + cjk) / totalNonWs >= threshold) return "Japanese";

  // Korean
  if (hangul / totalNonWs >= threshold) return "Korean";

  // Chinese: CJK without Kana or Hangul
  if (cjk / totalNonWs >= threshold && kana === 0 && hangul === 0) return "Chinese";

  // Other scripts
  const others = [
    ["CYRILLIC", "Russian"],
    ["ARABIC", "Arabic"],
    ["THAI", "Thai"],
    ["DEVANAGARI", "Hindi"]
  ];
  for (const [script, lang] of others) {
    if ((counts[script] || 0) / totalNonWs >= threshold) return lang;
  }

  return null;
}

export function languageGuidance(language) {
  return (
    `⚠️ LANGUAGE REQUIREMENT — NON-NEGOTIABLE ⚠️\n\n` +
    `The user wrote their prompt in ${language}. ` +
    `You MUST write your ENTIRE response in ${language}. ` +
    `Every word of output — title, mechanism, prose, trailing thread — ` +
    `must be in ${language}. Do NOT write in English. ` +
    `Only universally recognized proper nouns (brand names, place names) ` +
    `may remain in their original form. Everything else: ${language}.`
  );
}

File: src/pipeidea/language.py
"""Detect the dominant language of user seeds via Unicode script analysis."""

from __future__ import annotations

import unicodedata

# Unicode script ranges mapped to language names.
# We use the Unicode General Category + script lookup via unicodedata.
_SCRIPT_LANGUAGES: list[tuple[str, str, set[str]]] = [
    # (language_name, display_name, set of Unicode script names)
    ("Chinese", "Chinese", {"CJK"}),
    ("Japanese", "Japanese", {"HIRAGANA", "KATAKANA"}),
    ("Korean", "Korean", {"HANGUL"}),
    ("Russian", "Russian", {"CYRILLIC"}),
    ("Arabic", "Arabic", {"ARABIC"}),
    ("Thai", "Thai", {"THAI"}),
    ("Hindi", "Hindi", {"DEVANAGARI"}),
]


def _script_of(ch: str) -> str:
    """Return a simplified script category for a character."""
    name = unicodedata.name(ch, "")
    upper = name.upper()

    # CJK Unified Ideographs (shared by Chinese/Japanese/Korean)
    if "CJK" in upper:
        return "CJK"

    # Japanese-specific scripts
    if "HIRAGANA" in upper:
        return "HIRAGANA"
    if "KATAKANA" in upper:
        return "KATAKANA"

    # Korean
    if "HANGUL" in upper:
        return "HANGUL"

    # Cyrillic
    if "CYRILLIC" in upper:
        return "CYRILLIC"

    # Arabic
    if "ARABIC" in upper:
        return "ARABIC"

    # Thai
    if "THAI" in upper:
        return "THAI"

    # Devanagari
    if "DEVANAGARI" in upper:
        return "DEVANAGARI"

    return "OTHER"


def detect_seed_language(seeds: list[str]) -> str | None:
    """Detect the dominant non-English language from seed text.

    Returns a language name (e.g. "Chinese", "Japanese") if ≥30% of
    non-whitespace characters belong to a non-Latin script.
    Returns ``None`` for English / Latin-script input.
    """
    combined = " ".join(seeds).strip()
    if not combined:
        return None

    # Count characters by script
    script_counts: dict[str, int] = {}
    total_non_ws = 0

    for ch in combined:
        if ch.isspace():
            continue
        total_non_ws += 1
        script = _script_of(ch)
        script_counts[script] = script_counts.get(script, 0) + 1

    if total_non_ws == 0:
        return None

    # Check each language's scripts against the threshold
    threshold = 0.30

    # Special case: Japanese uses both CJK ideographs and Kana.
    # If any Kana is present, it's Japanese (CJK alone could be Chinese).
    kana_count = script_counts.get("HIRAGANA", 0) + script_counts.get("KATAKANA", 0)
    cjk_count = script_counts.get("CJK", 0)

    if kana_count > 0 and (kana_count + cjk_count) / total_non_ws >= threshold:
        return "Japanese"

    # Korean: Hangul characters present
    hangul_count = script_counts.get("HANGUL", 0)
    if hangul_count / total_non_ws >= threshold:
        return "Korean"

    # Chinese: CJK without Kana or Hangul
    if cjk_count / total_non_ws >= threshold and kana_count == 0 and hangul_count == 0:
        return "Chinese"

    # Other scripts
    for lang, _display, scripts in _SCRIPT_LANGUAGES:
        if lang in ("Chinese", "Japanese", "Korean"):
            continue  # Already handled above
        count = sum(script_counts.get(s, 0) for s in scripts)
        if count / total_non_ws >= threshold:
            return lang

    return None


def language_guidance(language: str) -> str:
    """Build the runtime guidance string for a detected language."""
    return (
        f"⚠️ LANGUAGE REQUIREMENT — NON-NEGOTIABLE ⚠️\n\n"
        f"The user wrote their prompt in {language}. "
        f"You MUST write your ENTIRE response in {language}. "
        f"Every word of output — title, mechanism, prose, trailing thread — "
        f"must be in {language}. Do NOT write in English. "
        f"Only universally recognized proper nouns (brand names, place names) "
        f"may remain in their original form. Everything else: {language}."
    )
